### ISSUE TITLE
参数image_directory不起作用

### ISSUE THREAD

**Issue by jinshuqishi2019** [Undertone0809/md-img-converter#8]
你好，修改图片储存文件夹不起作用，都是默认在 mds/converted_mds/images/ 希望能够修改，谢谢。

**Comment by Undertone0809** [comment]
Can you attach the relevant code？

**Comment by jinshuqishi2019** [comment]
感谢你的解答 
python 文件中的参数设置为 
` md_folder = MdFolder(name="md",image_directory="converted_mds/images")`
出现错误为


```
root@pdf:/md# python3 duo.py 
INFO:imarkdown.converter:[imarkdown] read file </md/md/5.md> successfully
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 58, in _download_img
    os.mkdir(image_local_storage_directory)
FileNotFoundError: [Errno 2] No such file or directory: 'md/converted_mds/images/'
ERROR:imarkdown.converter:[imarkdown] download_img failed, reason: [Errno 2] No such file or directory: 'md/converted_mds/images/'
Traceback (most recent call last):
  File "/md/duo.py", line 10, in <module>
    main()
  File "/md/duo.py", line 8, in main
```

还有一个问题就是下载cubox的图片出现图片损坏，不能打开，是不是格式判断错误了吧
![image](https://github.com/Undertone0809/imarkdown/assets/57516638/f862526c-1e25-4792-bb50-950b58733587)

下面是相应的md文件
[10.md](https://github.com/Undertone0809/imarkdown/files/12209193/10.md)

**Comment by Undertone0809** [comment]
Can you attach all code？I need to see more information.

**Comment by jinshuqishi2019** [comment]
这个duo.py 在 /md/duo.py
```
from imarkdown import LocalFileAdapter, MdFolder, MdImageConverter

def main():
    adapter = LocalFileAdapter()
    converter = MdImageConverter(adapter=adapter)
    
    md_folder = MdFolder(name="md",image_directory="converted_mds/images")
    converter.convert(md_folder, output_directory="converted_mds")
if __name__ == "__main__":
    main()
```
md文件在/md/md/*.md

现在根据错误的提示，需要创建/md/md/converted_mds/images/ 才能运行程序。

**Comment by Undertone0809** [comment]
Can you run this code and let me see the log terminal output?


```python
import logging
from imarkdown import LocalFileAdapter, MdFolder, MdImageConverter

def main():
    logging.basicConfig(level=logging.DEBUG)
    adapter = LocalFileAdapter()
    converter = MdImageConverter(adapter=adapter)
    
    md_folder = MdFolder(name="md",image_directory="converted_mds/images")
    converter.convert(md_folder, output_directory="converted_mds")

if __name__ == "__main__":
    main()
```

**Comment by Undertone0809** [comment]
> 这个duo.py 在 /md/duo.py
> 
> ```
> from imarkdown import LocalFileAdapter, MdFolder, MdImageConverter
> 
> def main():
>     adapter = LocalFileAdapter()
>     converter = MdImageConverter(adapter=adapter)
>     
>     md_folder = MdFolder(name="md",image_directory="converted_mds/images")
>     converter.convert(md_folder, output_directory="converted_mds")
> if __name__ == "__main__":
>     main()
> ```
> 
> md文件在/md/md/*.md
> 
> 现在根据错误的提示，需要创建/md/md/converted_mds/images/ 才能运行程序。

In Linux, you may need to grant read and write permissions to this file. Or you are mac?

**Comment by jinshuqishi2019** [comment]
我是使用linux，root用户，应该不存在权限问题
```
root@pdf:/md# cat test2.py 
import logging
from imarkdown import LocalFileAdapter, MdFolder, MdImageConverter

def main():
    logging.basicConfig(level=logging.DEBUG)
    adapter = LocalFileAdapter()
    converter = MdImageConverter(adapter=adapter)
    
    md_folder = MdFolder(name="md",image_directory="converted_mds/images")
    converter.convert(md_folder, output_directory="converted_mds")
if __name__ == "__main__":
    main()
root@pdf:/md# python3 test2.py 
DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
DEBUG:imarkdown.converter:[imarkdown] local default adapter <MdAdapterType.Local>
DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter initialization params: {'adapter': LocalFileAdapter(name='Local', storage_path_prefix='')}
DEBUG:imarkdown.schema:[imarkdown] MdFile convert params {'md_file_path': '/md/md/5.md', 'enable_rename': False, 'output_md_directory': '/md/converted_mds', 'image_local_storage_directory': '/md/converted_mds/images', 'is_local_images': False, 'enable_save_images': True}
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter not rename.
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_original_directory: /md/md
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_output_directory: /md/converted_mds
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter set image_local_storage_directory: /md/converted_mds/images
INFO:imarkdown.converter:[imarkdown] read file </md/md/5.md> successfully
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): image.cubox.pro:443
DEBUG:urllib3.connectionpool:https://image.cubox.pro:443 "GET /article/2023032414365949351/26491.jpg HTTP/1.1" 200 12534
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 58, in _download_img
    os.mkdir(image_local_storage_directory)
FileNotFoundError: [Errno 2] No such file or directory: 'md/converted_mds/images/'
ERROR:imarkdown.converter:[imarkdown] download_img failed, reason: [Errno 2] No such file or directory: 'md/converted_mds/images/'
Traceback (most recent call last):
  File "/md/test2.py", line 12, in <module>
    main()
  File "/md/test2.py", line 10, in main
    converter.convert(md_folder, output_directory="converted_mds")
  File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 320, in convert
    self.converter.convert(**kwargs)
  File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 205, in convert
    modified_data: str = self._find_img_and_replace(original_data)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 233, in _find_img_and_replace
    converted_image_url = self._get_converted_image_url(image)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 257, in _get_converted_image_url
    raise Exception("get a empty image path")
Exception: get a empty image path
```

**Comment by Undertone0809** [comment]
> 

A error  occured in the regular expression. Your image url is not working properly it occipies `()`. I'm tring to find a new regular expression to solve this problem.

```
![](https://cubox.pro/c/filters:no_upscale()?valid=false&imageUrl=https%3A%2F%2Fpicx.zhimg.com%2F50%2Fv2-132e174423fc0785a386ff534e824ebc_720w.jpg%3Fsource%3D1940ef5c)
```

**Comment by Undertone0809** [comment]
> 我是使用linux，root用户，应该不存在权限问题
> 
> ```
> root@pdf:/md# cat test2.py 
> import logging
> from imarkdown import LocalFileAdapter, MdFolder, MdImageConverter
> 
> def main():
>     logging.basicConfig(level=logging.DEBUG)
>     adapter = LocalFileAdapter()
>     converter = MdImageConverter(adapter=adapter)
>     
>     md_folder = MdFolder(name="md",image_directory="converted_mds/images")
>     converter.convert(md_folder, output_directory="converted_mds")
> if __name__ == "__main__":
>     main()
> root@pdf:/md# python3 test2.py 
> DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
> DEBUG:imarkdown.converter:[imarkdown] local default adapter <MdAdapterType.Local>
> DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter initialization params: {'adapter': LocalFileAdapter(name='Local', storage_path_prefix='')}
> DEBUG:imarkdown.schema:[imarkdown] MdFile convert params {'md_file_path': '/md/md/5.md', 'enable_rename': False, 'output_md_directory': '/md/converted_mds', 'image_local_storage_directory': '/md/converted_mds/images', 'is_local_images': False, 'enable_save_images': True}
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter not rename.
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_original_directory: /md/md
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_output_directory: /md/converted_mds
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter set image_local_storage_directory: /md/converted_mds/images
> INFO:imarkdown.converter:[imarkdown] read file </md/md/5.md> successfully
> DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): image.cubox.pro:443
> DEBUG:urllib3.connectionpool:https://image.cubox.pro:443 "GET /article/2023032414365949351/26491.jpg HTTP/1.1" 200 12534
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 58, in _download_img
>     os.mkdir(image_local_storage_directory)
> FileNotFoundError: [Errno 2] No such file or directory: 'md/converted_mds/images/'
> ERROR:imarkdown.converter:[imarkdown] download_img failed, reason: [Errno 2] No such file or directory: 'md/converted_mds/images/'
> Traceback (most recent call last):
>   File "/md/test2.py", line 12, in <module>
>     main()
>   File "/md/test2.py", line 10, in main
>     converter.convert(md_folder, output_directory="converted_mds")
>   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 320, in convert
>     self.converter.convert(**kwargs)
>   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 205, in convert
>     modified_data: str = self._find_img_and_replace(original_data)
>                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 233, in _find_img_and_replace
>     converted_image_url = self._get_converted_image_url(image)
>                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 257, in _get_converted_image_url
>     raise Exception("get a empty image path")
> Exception: get a empty image path
> ```


I know the problem. The problem occured in `os.mkdir()`.


"`os.mkdir()` and `os.makedirs()` are both built-in functions in Python used to create directories in a file system. Here are the differences between them:

1. `os.mkdir(path, mode=0o777, *, dir_fd=None)`:
   - Creates a single directory.
   - `path` is the path of the directory to be created.
   - `mode` is the permission of the directory, with a default value of 0o777 (octal).
   - `dir_fd` is an optional file descriptor parameter used to specify the directory to create the directory in, based on the file descriptor.
   - If any intermediate directories in the specified path do not exist, a `FileNotFoundError` exception is raised.

2. `os.makedirs(name, mode=0o777, exist_ok=False)`:
   - Creates multiple directories, including any intermediate directories that do not exist.
   - `name` is the path of the directory to be created.
   - `mode` is the permission of the directory, with a default value of 0o777 (octal).
   - `exist_ok` is a boolean value. If set to `True`, no exception is raised when the directory already exists. If set to `False`, a `FileExistsError` exception is raised if the directory already exists.

In summary, `os.mkdir()` can only create a single directory, while `os.makedirs()` can create multiple directories. If there are intermediate directories that do not exist in the desired path, `os.makedirs()` should be used. If you only need to create a single directory or if you are certain that the intermediate directories already exist, you can use `os.mkdir()`."

**Comment by Undertone0809** [comment]
> > 我是使用linux，root用户，应该不存在权限问题
> > ```
> > root@pdf:/md# cat test2.py 
> > import logging
> > from imarkdown import LocalFileAdapter, MdFolder, MdImageConverter
> > 
> > def main():
> >     logging.basicConfig(level=logging.DEBUG)
> >     adapter = LocalFileAdapter()
> >     converter = MdImageConverter(adapter=adapter)
> >     
> >     md_folder = MdFolder(name="md",image_directory="converted_mds/images")
> >     converter.convert(md_folder, output_directory="converted_mds")
> > if __name__ == "__main__":
> >     main()
> > root@pdf:/md# python3 test2.py 
> > DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
> > DEBUG:imarkdown.converter:[imarkdown] local default adapter <MdAdapterType.Local>
> > DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
> > DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter initialization params: {'adapter': LocalFileAdapter(name='Local', storage_path_prefix='')}
> > DEBUG:imarkdown.schema:[imarkdown] MdFile convert params {'md_file_path': '/md/md/5.md', 'enable_rename': False, 'output_md_directory': '/md/converted_mds', 'image_local_storage_directory': '/md/converted_mds/images', 'is_local_images': False, 'enable_save_images': True}
> > DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter not rename.
> > DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_original_directory: /md/md
> > DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_output_directory: /md/converted_mds
> > DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter set image_local_storage_directory: /md/converted_mds/images
> > INFO:imarkdown.converter:[imarkdown] read file </md/md/5.md> successfully
> > DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): image.cubox.pro:443
> > DEBUG:urllib3.connectionpool:https://image.cubox.pro:443 "GET /article/2023032414365949351/26491.jpg HTTP/1.1" 200 12534
> > Traceback (most recent call last):
> >   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 58, in _download_img
> >     os.mkdir(image_local_storage_directory)
> > FileNotFoundError: [Errno 2] No such file or directory: 'md/converted_mds/images/'
> > ERROR:imarkdown.converter:[imarkdown] download_img failed, reason: [Errno 2] No such file or directory: 'md/converted_mds/images/'
> > Traceback (most recent call last):
> >   File "/md/test2.py", line 12, in <module>
> >     main()
> >   File "/md/test2.py", line 10, in main
> >     converter.convert(md_folder, output_directory="converted_mds")
> >   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 320, in convert
> >     self.converter.convert(**kwargs)
> >   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 205, in convert
> >     modified_data: str = self._find_img_and_replace(original_data)
> >                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 233, in _find_img_and_replace
> >     converted_image_url = self._get_converted_image_url(image)
> >                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >   File "/usr/local/lib/python3.11/dist-packages/imarkdown/converter.py", line 257, in _get_converted_image_url
> >     raise Exception("get a empty image path")
> > Exception: get a empty image path
> > ```
> 
> I know the problem. The problem occured in `os.mkdir()`.
> 
> "`os.mkdir()` and `os.makedirs()` are both built-in functions in Python used to create directories in a file system. Here are the differences between them:
> 
> 1. `os.mkdir(path, mode=0o777, *, dir_fd=None)`:
>    
>    * Creates a single directory.
>    * `path` is the path of the directory to be created.
>    * `mode` is the permission of the directory, with a default value of 0o777 (octal).
>    * `dir_fd` is an optional file descriptor parameter used to specify the directory to create the directory in, based on the file descriptor.
>    * If any intermediate directories in the specified path do not exist, a `FileNotFoundError` exception is raised.
> 2. `os.makedirs(name, mode=0o777, exist_ok=False)`:
>    
>    * Creates multiple directories, including any intermediate directories that do not exist.
>    * `name` is the path of the directory to be created.
>    * `mode` is the permission of the directory, with a default value of 0o777 (octal).
>    * `exist_ok` is a boolean value. If set to `True`, no exception is raised when the directory already exists. If set to `False`, a `FileExistsError` exception is raised if the directory already exists.
> 
> In summary, `os.mkdir()` can only create a single directory, while `os.makedirs()` can create multiple directories. If there are intermediate directories that do not exist in the desired path, `os.makedirs()` should be used. If you only need to create a single directory or if you are certain that the intermediate directories already exist, you can use `os.mkdir()`."

This bug has fix in #9 . You can use it by `pip install -U imarkdown` or `git clone https://github.com/Undertone0809/imarkdown.git`

Regular expression is still a problem, my re is `r"(?:!\[(.*?)\]\((.*?)\))|<img.*?src=[\'\"](.*?)[\'\"].*?>"`. Do you have some good idea to solve it?

**Comment by jinshuqishi2019** [comment]
已经更新最新的版本，可以自动创建文件夹，但是指定图片文件夹还是不生效。
test2.py  中 已经写了 image_directory="converted_mds/images" 但是输出还是不生效
文件目录
```
root@pdf:/md# ls converted_mds 
1.md  10.md  11.md  2.md  3.md  4.md  5.md  6.md  7.md  8.md  9.md  images
root@pdf:/md# ls converted_mds/images/
root@pdf:/md# ls md/converted_mds/images/
20230731_0241482973.png  20230731_0241507355.png  20230731_0241552573.png  20230731_0242005304.png  20230731_0242047722.png  20230731_0242099147.png
20230731_0241483250.png  20230731_0241511162.png  20230731_0241555772.png  20230731_0242007537.png  
```


输出代码
```
root@pdf:/md# cat test2.py 
import logging
from imarkdown import LocalFileAdapter, MdFolder, MdImageConverter

def main():
    logging.basicConfig(level=logging.DEBUG)
    adapter = LocalFileAdapter()
    converter = MdImageConverter(adapter=adapter)
    
    md_folder = MdFolder(name="md",image_directory="converted_mds/images")
    converter.convert(md_folder, output_directory="converted_mds")
if __name__ == "__main__":
    main()
root@pdf:/md# rm -r converted_mds/
root@pdf:/md# rm -r md/converted_mds/images/
root@pdf:/md# ls
duo.py  md  test.py  test2.py
root@pdf:/md# python3 test2.py 
DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
DEBUG:imarkdown.converter:[imarkdown] local default adapter <MdAdapterType.Local>
DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter initialization params: {'adapter': LocalFileAdapter(name='Local', storage_path_prefix='')}
DEBUG:imarkdown.schema:[imarkdown] MdFile convert params {'md_file_path': '/md/md/5.md', 'enable_rename': False, 'output_md_directory': '/md/converted_mds', 'image_local_storage_directory': '/md/converted_mds/images', 'is_local_images': False, 'enable_save_images': True}
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter not rename.
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_original_directory: /md/md
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_output_directory: /md/converted_mds
DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter set image_local_storage_directory: /md/converted_mds/images
INFO:imarkdown.converter:[imarkdown] read file </md/md/5.md> successfully
```

**Comment by Undertone0809** [comment]
> 已经更新最新的版本，可以自动创建文件夹，但是指定图片文件夹还是不生效。 test2.py 中 已经写了 image_directory="converted_mds/images" 但是输出还是不生效 文件目录
> 
> ```
> root@pdf:/md# ls converted_mds 
> 1.md  10.md  11.md  2.md  3.md  4.md  5.md  6.md  7.md  8.md  9.md  images
> root@pdf:/md# ls converted_mds/images/
> root@pdf:/md# ls md/converted_mds/images/
> 20230731_0241482973.png  20230731_0241507355.png  20230731_0241552573.png  20230731_0242005304.png  20230731_0242047722.png  20230731_0242099147.png
> 20230731_0241483250.png  20230731_0241511162.png  20230731_0241555772.png  20230731_0242007537.png  
> ```
> 
> 输出代码
> 
> ```
> root@pdf:/md# cat test2.py 
> import logging
> from imarkdown import LocalFileAdapter, MdFolder, MdImageConverter
> 
> def main():
>     logging.basicConfig(level=logging.DEBUG)
>     adapter = LocalFileAdapter()
>     converter = MdImageConverter(adapter=adapter)
>     
>     md_folder = MdFolder(name="md",image_directory="converted_mds/images")
>     converter.convert(md_folder, output_directory="converted_mds")
> if __name__ == "__main__":
>     main()
> root@pdf:/md# rm -r converted_mds/
> root@pdf:/md# rm -r md/converted_mds/images/
> root@pdf:/md# ls
> duo.py  md  test.py  test2.py
> root@pdf:/md# python3 test2.py 
> DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
> DEBUG:imarkdown.converter:[imarkdown] local default adapter <MdAdapterType.Local>
> DEBUG:imarkdown.config:[imarkdown] get last adapter <Local>
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter initialization params: {'adapter': LocalFileAdapter(name='Local', storage_path_prefix='')}
> DEBUG:imarkdown.schema:[imarkdown] MdFile convert params {'md_file_path': '/md/md/5.md', 'enable_rename': False, 'output_md_directory': '/md/converted_mds', 'image_local_storage_directory': '/md/converted_mds/images', 'is_local_images': False, 'enable_save_images': True}
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter not rename.
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_original_directory: /md/md
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter md_file_output_directory: /md/converted_mds
> DEBUG:imarkdown.converter:[imarkdown] BaseMdImageConverter set image_local_storage_directory: /md/converted_mds/images
> INFO:imarkdown.converter:[imarkdown] read file </md/md/5.md> successfully
> ```

I have tried your code and I didn'f fail. You can delete original converted file and retry. Moreover, your image is a bit different. You can use custom regular expression in v1.2.1.  #10 [Documentation](https://github.com/Undertone0809/imarkdown/blob/main/README_zh.md#%E8%87%AA%E5%AE%9A%E4%B9%89%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F) Hoping this helps.

**Comment by jinshuqishi2019** [comment]
尝试了许多遍，即使重新配置了linux，降低python到3.8.0版本，参数image_directory还是不起作用，只能后期使用脚本将地址批量修改。
还有你增加的 re_rule 参数，但是相应的converter.py中还是固定了 image[1]，不是很自由，或许循环一下找到http更好。

**Comment by jinshuqishi2019** [comment]
还有一个建议，就是markdown的图片失效了，现在是出现异常停止了，建议可以搞成跳过失效的图片，然后结束的时候输出下载失败的地方。

**Comment by Undertone0809** [comment]
> 尝试了许多遍，即使重新配置了linux，降低python到3.8.0版本，参数image_directory还是不起作用，只能后期使用脚本将地址批量修改。 还有你增加的 re_rule 参数，但是相应的converter.py中还是固定了 image[1]，不是很自由，或许循环一下找到http更好。

Thank you for your suggestion. I have been update the `re_rule` to the `ElementFinder` in order to provide more custom flexibility at `v1.2.2`. You can see [documentation](https://github.com/Undertone0809/imarkdown/blob/main/README_zh.md#%E8%87%AA%E5%AE%9A%E4%B9%89%E5%85%83%E7%B4%A0%E6%9F%A5%E6%89%BE%E5%99%A8).

**Comment by Undertone0809** [comment]
> 常停止了，建议可以搞成跳过

Thank you for your suggestion. I will provide more feature later.

**Comment by Undertone0809** [comment]
> 还有一个建议，就是markdown的图片失效了，现在是出现异常停止了，建议可以搞成跳过失效的图片，然后结束的时候输出下载失败的地方。

Thank you for your suggestion. I will provide more feature later.

**Comment by jinshuqishi2019** [comment]
感谢你的帮助！

**Comment by jinshuqishi2019** [comment]
我找到一段代码可以不使用正则寻找图片地址，而且适用于markdown不同的图片格式。你可以参考一下。
```
import markdown
from bs4 import BeautifulSoup

with open("md/1.md", "r") as f:
    # 读取文件内容
    content = f.read()

# 将 Markdown 文本转换为 HTML 文本
html = markdown.markdown(content)

# 使用 BeautifulSoup 库解析 HTML 文本
soup = BeautifulSoup(html, 'html.parser')

# 提取图片链接
images = soup.find_all('img')

# 遍历列表并打印出每个图片链接
for image in images:
    print(image['src'])
```

**Comment by Undertone0809** [comment]
It looks better than regular expression, thank you!
